### PR TITLE
feat(campaign): allows to burn vouchers

### LINKED ISSUE #268: campaign: allows to burn vouchers
Implement `MsgBurnVouchers` that allows to burn owned vouchers and decrease `AllocatedShares` for the associated shares

### PR DESCRIPTION
**Pull request by Pantani**
Closes #268

### What does this PR does?

Implement `MsgBurnVouchers` to burn vouchers and decrease the campaign AllocatedShares.

### CODE REVIEW

File: proto/campaign/tx.proto [hunk 1]
```diff
@@ -100,4 +101,13 @@ message MsgMintVouchers {
 message MsgMintVouchersResponse {
 }
 
+message MsgBurnVouchers {
+  string creator = 1;
```
Comment (lubtd) [inline proto/campaign/tx.proto:105]:
```suggestion
  string sender = 1;
```

File: x/campaign/client/cli/tx_burn_vouchers.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/tendermint/spn/x/campaign/types"
+)
+
+var _ = strconv.Itoa(0)
+
+func CmdBurnVouchers() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "burn-vouchers [campaign-id] [vouchers]",
+		Short: "Burn vouchers",
```
Comment (lubtd) [inline x/campaign/client/cli/tx_burn_vouchers.go:21]:
```suggestion
		Short: "Burn vouchers and decrease allocated shares of the campaign",
```

File: x/campaign/keeper/msg_server_burn_vouchers.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	}
+
+	creator, err := sdk.AccAddressFromBech32(msg.Creator)
+	if err != nil {
+		return nil, spnerrors.Criticalf("can't parse creator address %s", err.Error())
+	}
+
+	// transfer vouchers
+	if err := k.bankKeeper.SendCoinsFromAccountToModule(ctx, creator, types.ModuleName, msg.Vouchers); err != nil {
+		return nil, spnerrors.Criticalf("can't send burned coins %s", err.Error())
```
Comment (lubtd) [inline x/campaign/keeper/msg_server_burn_vouchers.go:27]: This is not a critical error because the message signer balance has not been checked previously and can't be statically verified

File: x/campaign/keeper/msg_server_burn_vouchers.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+	}
+
+	// Burn vouchers
+	if err := k.bankKeeper.BurnCoins(ctx, types.ModuleName, msg.Vouchers); err != nil {
+		return nil, sdkerrors.Wrap(types.ErrVouchersBurn, err.Error())
```
Comment (lubtd) [inline x/campaign/keeper/msg_server_burn_vouchers.go:32]: Here this is a critical error since the module is supposed to have received the funds

File: x/campaign/keeper/msg_server_burn_vouchers.go [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+	shares, err := types.VouchersToShares(msg.Vouchers, msg.CampaignID)
+	if err != nil {
+		return nil, spnerrors.Criticalf("verified voucher are invalid %s", err.Error())
+	}
+
+	// Decrease the campaign shares
+	campaign.AllocatedShares, err = types.DecreaseShares(campaign.AllocatedShares, shares)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrInvalidShares, "%d", msg.CampaignID)
```
Comment (lubtd) [inline x/campaign/keeper/msg_server_burn_vouchers.go:43]: This is also a critical error since the shares are valid they are supposed to be previously allocated for the campaign

File: x/campaign/keeper/msg_server_burn_vouchers_test.go [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+			name: "should not burn more than allocated shares",
+			msg: types.MsgBurnVouchers{
+				Creator:    addr.String(),
+				CampaignID: campaign.Id,
+				Vouchers:   vouchersTooBig,
+			},
+			err: spnerrors.ErrCritical,
+		},
+		{
+			name: "burn vouchers",
```
Comment (lubtd) [inline x/campaign/keeper/msg_server_burn_vouchers_test.go:76]: We should test this message several time for the same campaign and test the edge case once all vouchers are burned

File: x/campaign/types/errors.go [hunk 7]
```diff
@@ -18,4 +18,6 @@ var (
 	ErrAccountNotFound     = sdkerrors.Register(ModuleName, 8, "account not found")
 	ErrSharesDecrease      = sdkerrors.Register(ModuleName, 9, "shares can't be decreased")
 	ErrVouchersMinting     = sdkerrors.Register(ModuleName, 10, "vouchers can't be minted")
+	ErrInvalidVouchers     = sdkerrors.Register(ModuleName, 11, "invalid vouchers")
+	ErrVouchersBurn        = sdkerrors.Register(ModuleName, 12, "vouchers can't be burned")
```
Comment (lubtd) [inline x/campaign/types/errors.go:22]: Not needed, the bank's burn method should always succeed since funds are first sent to the module account. So handled with a critical error

File: x/campaign/types/message_burn_vouchers.go [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+	}
+	return []sdk.AccAddress{creator}
+}
+
+func (msg *MsgBurnVouchers) GetSignBytes() []byte {
+	bz := ModuleCdc.MustMarshalJSON(msg)
+	return sdk.MustSortJSON(bz)
+}
+
+func (msg *MsgBurnVouchers) ValidateBasic() error {
```
Comment (lubtd) [inline x/campaign/types/message_burn_vouchers.go:39]: Since we have the campaign ID in the message and the campaign ID in voucher's denom, we should statically check that all vouchers in the messages are linked to the provided campaign

File: x/campaign/types/message_burn_vouchers_test.go [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+			name: "invalid address",
+			msg: types.MsgBurnVouchers{
+				Creator:    "invalid_address",
+				CampaignID: 0,
+				Vouchers:   sample.Coins(),
+			},
+			err: sdkerrors.ErrInvalidAddress,
+		},
+		{
+			name: "valid address",
```
Comment (lubtd) [inline x/campaign/types/message_burn_vouchers_test.go:29]:
```suggestion
			name: "valid message",
```
We can also move this one first

File: x/campaign/types/message_burn_vouchers_test.go [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+		name string
+		msg  types.MsgBurnVouchers
+		err  error
+	}{
+		{
+			name: "invalid address",
+			msg: types.MsgBurnVouchers{
+				Creator:    "invalid_address",
+				CampaignID: 0,
+				Vouchers:   sample.Coins(),
```
Comment (lubtd) [inline x/campaign/types/message_burn_vouchers_test.go:24]:
```suggestion
				Vouchers:   sample.Coins(),
```

We should implement `sample.Voucher(campaignID)` so we can create vouchers linked to the campaign

**Review by lubtd — CHANGES_REQUESTED** [review]

File: x/campaign/keeper/msg_server_burn_vouchers_test.go (hunk 6 above)
Comment (Pantani) [reply x/campaign/keeper/msg_server_burn_vouchers_test.go:76]: we already have a [test for all burned vouchers](https://github.com/tendermint/spn/pull/329/files#diff-dc5083900b9c760a253d327bac3c61ab00573fe77a5ba660fa5d733e9f127dbaR84-R90), this `burn vouchers` test, burn all vouchers, and the seconds try to burn again, and it's will fails

Comment (lubtd) [reply x/campaign/keeper/msg_server_burn_vouchers_test.go:76]: I meant we should burn some and then try to burn again on the same campaign

File: x/campaign/types/errors.go [hunk 11]
```diff
@@ -18,4 +18,6 @@ var (
 	ErrAccountNotFound     = sdkerrors.Register(ModuleName, 8, "account not found")
 	ErrSharesDecrease      = sdkerrors.Register(ModuleName, 9, "shares can't be decreased")
 	ErrVouchersMinting     = sdkerrors.Register(ModuleName, 10, "vouchers can't be minted")
+	ErrInvalidVouchers     = sdkerrors.Register(ModuleName, 11, "invalid vouchers")
+	ErrInsufficientFunds   = sdkerrors.Register(ModuleName, 12, "account with insufficient funds")
```
Comment (lubtd) [inline x/campaign/types/errors.go:22]: We can precise vouchers here

File: x/campaign/types/message_burn_vouchers_test.go [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+			name: "invalid shares",
+			msg: types.MsgBurnVouchers{
+				Sender:     sample.AccAddress(),
+				CampaignID: 0,
+				Vouchers:   invalidCoins,
+			},
+			err: types.ErrInvalidVouchers,
+		},
+		{
+			name: "empty shares",
```
Comment (lubtd) [inline x/campaign/types/message_burn_vouchers_test.go:46]:
```suggestion
			name: "empty vouchers",
```

File: x/campaign/types/message_burn_vouchers_test.go [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+		{
+			name: "valid message",
+			msg: types.MsgBurnVouchers{
+				Sender:     sample.AccAddress(),
+				CampaignID: 0,
+				Vouchers:   sample.Vouchers(0),
+			},
+		},
+		{
+			name: "invalid shares",
```
Comment (lubtd) [inline x/campaign/types/message_burn_vouchers_test.go:37]:
```suggestion
			name: "invalid vouchers",
```

File: x/campaign/types/message_burn_vouchers_test.go [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+			name: "empty shares",
+			msg: types.MsgBurnVouchers{
+				Sender:     sample.AccAddress(),
+				CampaignID: 0,
+				Vouchers:   sdk.Coins{},
+			},
+			err: types.ErrInvalidVouchers,
+		},
+		{
+			name: "invalid vouchers",
```
Comment (lubtd) [inline x/campaign/types/message_burn_vouchers_test.go:55]:
```suggestion
			name: "vouchers not from the associated campaign",
```

File: x/campaign/keeper/msg_server_burn_vouchers_test.go (hunk 6 above)
Comment (Pantani) [reply x/campaign/keeper/msg_server_burn_vouchers_test.go]: added on https://github.com/tendermint/spn/pull/329/commits/7dea6680a2e389986c22ae538b1c7051cf19b24d

File: x/campaign/types/errors.go (hunk 11 above)
Comment (Pantani) [reply x/campaign/types/errors.go:22]: Sorry, I didn't understand. Do you mean the error description?

Comment (lubtd) [reply x/campaign/types/errors.go]: yes

**Review by lubtd — APPROVED** [review]
👍
